File: tests/package-management/python-tests/package_service_tests.py
import json
import os
import tempfile
import uuid
from typing import cast

from apollo3zehn_package_management import PackageReference, PackageService
from apollo3zehn_package_management._encoder import JsonEncoder


async def can_create_package_reference_test():
    
    # Arrange
    with tempfile.NamedTemporaryFile(delete=True) as temp_file:
        config_folder_path = temp_file.name
        
    package_service = PackageService(config_folder_path)

    package_reference = PackageReference(
        provider="foo",
        configuration=cast(dict[str, str], None)
    )

    # Act
    expected_id = await package_service.put(package_reference)

    # Assert
    with open(os.path.join(config_folder_path, "packages.json"), "r") as file:
        actual_package_reference_map = JsonEncoder.decode(dict[uuid.UUID, PackageReference], json.load(file))
        
    entry = list(actual_package_reference_map.items())[0]

    assert entry[0] == expected_id
    assert entry[1].provider == "foo"
    assert entry[1].configuration is None

async def can_get_package_reference_test():

    # Arrange
    id1 = uuid.uuid4()
    id2 = uuid.uuid4()

    package_reference_map = {
        id1: PackageReference(
            provider="foo",
            configuration=cast(dict[str, str], None)
        ),
        id2: PackageReference(
            provider="bar",
            configuration=cast(dict[str, str], None)
        )
    }

    json_value = JsonEncoder.encode(package_reference_map)

    with tempfile.NamedTemporaryFile(delete=True) as temp_file:
        config_folder_path = temp_file.name

    os.makedirs(config_folder_path)

    with open(os.path.join(config_folder_path, "packages.json"), "w") as file:
        json.dump(json_value, file)

    package_service = PackageService(config_folder_path)

    # Act
    actual = await package_service.get(id2)

    # Assert
    assert json.dumps(JsonEncoder.encode(actual)) == \
        json.dumps(JsonEncoder.encode(package_reference_map[id2]))
    
async def can_try_update_package_reference_test():

    # Arrange
    id1 = uuid.uuid4()
    id2 = uuid.uuid4()

    package_reference_map = {
        id1: PackageReference(
            provider="foo",
            configuration=cast(dict[str, str], None)
        ),
        id2: PackageReference(
            provider="bar",
            configuration=cast(dict[str, str], None)
        )
    }

    json_value = JsonEncoder.encode(package_reference_map)

    with tempfile.NamedTemporaryFile(delete=True) as temp_file:
        config_folder_path = temp_file.name

    os.makedirs(config_folder_path)

    with open(os.path.join(config_folder_path, "packages.json"), "w") as file:
        json.dump(json_value, file)
        
    package_service = PackageService(config_folder_path)

    new_package_reference = PackageReference(
        provider="baz",
        configuration=cast(dict[str, str], None)
    )

    expected = package_reference_map.copy()
    expected[id1] = new_package_reference

    # Act
    success = await package_service.try_update(id1, new_package_reference)

    # Assert
    assert success

    with open(os.path.join(config_folder_path, "packages.json")) as json_file:
        actual = json.load(json_file)

    assert json.dumps(JsonEncoder.encode(actual)) == \
        json.dumps(JsonEncoder.encode(expected))

async def can_delete_package_test():

    # Arrange
    id1 = uuid.uuid4()
    id2 = uuid.uuid4()

    package_reference_map = {
        id1: PackageReference(
            provider="foo",
            configuration=cast(dict[str, str], None)
        ),
        id2: PackageReference(
            provider="bar",
            configuration=cast(dict[str, str], None)
        )
    }

    json_value = JsonEncoder.encode(package_reference_map)

    with tempfile.NamedTemporaryFile(delete=True) as temp_file:
        config_folder_path = temp_file.name

    os.makedirs(config_folder_path)

    packages_file_path = os.path.join(config_folder_path, "packages.json")

    with open(packages_file_path, "w") as file:
        json.dump(json_value, file)

    package_service = PackageService(config_folder_path)

    # Act
    await package_service.delete(id1)

    # Assert
    with open(packages_file_path, "r") as file:
        actual_package_reference_map = json.load(file)

    assert str(id1) not in actual_package_reference_map
    assert str(id2) in actual_package_reference_map

async def can_get_all_packages_test():

    # Arrange
    id1 = uuid.uuid4()
    id2 = uuid.uuid4()

    package_reference_map = {
        id1: PackageReference(
            provider="foo",
            configuration=cast(dict[str, str], None)
        ),
        id2: PackageReference(
            provider="bar",
            configuration=cast(dict[str, str], None)
        )
    }

    json_value = JsonEncoder.encode(package_reference_map)

    with tempfile.NamedTemporaryFile(delete=True) as temp_file:
        config_folder_path = temp_file.name

    os.makedirs(config_folder_path)

    packages_file_path = os.path.join(config_folder_path, "packages.json")

    with open(packages_file_path, "w") as file:
        json.dump(json_value, file)

    package_service = PackageService(config_folder_path)

    # Act
    actual_package_map = await package_service.get_all()

    # Assert
    expected = json.dumps(JsonEncoder.encode(package_reference_map))
    actual = json.dumps(JsonEncoder.encode(actual_package_map))

    assert actual == expected

File: tests/package-management/dotnet-tests/PackageServiceTests.cs
// MIT License
// Copyright (c) [2024] [Apollo3zehn]

using Apollo3zehn.PackageManagement;
using Apollo3zehn.PackageManagement.Core;
using Apollo3zehn.PackageManagement.Services;
using Microsoft.Extensions.Logging;
using Moq;
using System.Text.Json;
using Xunit;

namespace Services;

public class PackageServiceTests
{
    delegate bool GobbleReturns(out string? pipelineMap);

    [Fact]
    public async Task CanCreatePackageReference()
    {
        // Arrange
        var filePath = Path.GetTempFileName();
        var packageService = GetPackageService(filePath, []);

        var packageReference = new PackageReference(
            Provider: "foo",
            Configuration: default!
        );

        // Act
        var expectedId = await packageService.PutAsync(packageReference);

        // Assert
        var jsonString = File.ReadAllText(filePath);

        var actualPackageReferenceMap = JsonSerializer.Deserialize<Dictionary<Guid, PackageReference>>(
            jsonString,
            JsonSerializerOptions.Web
        )!;

        var entry = Assert.Single(actualPackageReferenceMap);

        Assert.Equal(expectedId, entry.Key);
        Assert.Equal("foo", entry.Value.Provider);
        Assert.Null(entry.Value.Configuration);
    }

    [Fact]
    public async Task CanGetPackageReference()
    {
        // Arrange
        var id1 = Guid.NewGuid();
        var id2 = Guid.NewGuid();

        var packageReferenceMap = new Dictionary<Guid, PackageReference>()
        {
            [id1] = new PackageReference(
                Provider: "foo",
                Configuration: default!
            ),
            [id2] = new PackageReference(
                Provider: "bar",
                Configuration: default!
            )
        };

        var packageService = GetPackageService(default!, packageReferenceMap);

        // Act
        var actual = await packageService.GetAsync(id2);

        // Assert
        Assert.Equal(
            expected: JsonSerializer.Serialize(actual),
            actual: JsonSerializer.Serialize(packageReferenceMap[id2])
        );
    }

    [Fact]
    public async Task CanTryUpdatePackageReference()
    {
        // Arrange
        var id1 = Guid.NewGuid();
        var id2 = Guid.NewGuid();

        var packageReferenceMap = new Dictionary<Guid, PackageReference>()
        {
            [id1] = new PackageReference(
                Provider: "foo",
                Configuration: default!
            ),
            [id2] = new PackageReference(
                Provider: "bar",
                Configuration: default!
            )
        };

        var filePath = Path.GetTempFileName();
        var packageService = GetPackageService(filePath, packageReferenceMap);

        var newPackageReference = new PackageReference(
            Provider: "baz",
            Configuration: default!
        );

        var expected = packageReferenceMap
            .ToDictionary(x => x.Key, x => x.Value);

        expected[id1] = newPackageReference;

        // Act
        var success = await packageService.TryUpdateAsync(id1, newPackageReference);

        // Assert
        Assert.True(success);

        var actual = JsonSerializer.Deserialize<Dictionary<Guid, PackageReference>>(
            File.ReadAllText(filePath),
            JsonSerializerOptions.Web
        );

        Assert.Equivalent(expected, actual, strict: true);
    }

    [Fact]
    public async Task CanDeletePackage()
    {
        // Arrange
        var id1 = Guid.NewGuid();
        var id2 = Guid.NewGuid();

        var packageReferencesMap = new Dictionary<Guid, PackageReference>()
        {
            [id1] = new PackageReference(
                Provider: "foo",
                Configuration: default!
            ),
            [id2] = new PackageReference(
                Provider: "bar",
                Configuration: default!
            )
        };

        var filePath = Path.GetTempFileName();
        var packageService = GetPackageService(filePath, packageReferencesMap);

        // Act
        await packageService.DeleteAsync(id1);

        // Assert
        packageReferencesMap.Remove(id1);
        var expected = JsonSerializerHelper.SerializeIndented(packageReferencesMap);
        var actual = File.ReadAllText(filePath);

        Assert.Equal(expected, actual);
    }

    [Fact]
    public async Task CanGetAllPackages()
    {
        // Arrange
        var id1 = Guid.NewGuid();
        var id2 = Guid.NewGuid();

        var packageReferencesMap = new Dictionary<Guid, PackageReference>()
        {
            [id1] = new PackageReference(
                Provider: "foo",
                Configuration: default!
            ),
            [id2] = new PackageReference(
                Provider: "bar",
                Configuration: default!
            )
        };

        var filePath = Path.GetTempFileName();
        var packageService = GetPackageService(filePath, packageReferencesMap);

        // Act
        var actualPackageMap = await packageService.GetAllAsync();

        // Assert
        var expected = JsonSerializerHelper.SerializeIndented(packageReferencesMap.OrderBy(current => current.Key));
        var actual = JsonSerializerHelper.SerializeIndented(actualPackageMap.OrderBy(current => current.Key));

        Assert.Equal(expected, actual);
    }

    private static IPackageService GetPackageService(
        string filePath,
        Dictionary<Guid, PackageReference> packageReferenceMap)
    {
        var databaseService = Mock.Of<IPackageManagementDatabaseService>();

        Mock.Get(databaseService)
            .Setup(databaseService => databaseService.TryReadPackageReferenceMap(out It.Ref<string?>.IsAny))
            .Returns(new GobbleReturns((out string? packageReferenceMapString) =>
            {
                packageReferenceMapString = JsonSerializer.Serialize(packageReferenceMap);
                return true;
            }));

        Mock.Get(databaseService)
            .Setup(databaseService => databaseService.WritePackageReferenceMap())
            .Returns(() => File.OpenWrite(filePath));

        var loggerFactory = new LoggerFactory();
        var packageService = new PackageService(databaseService, loggerFactory);

        return packageService;
    }
}